packages/vite/src/streamHelpers.ts

```diff
@@ -0,0 +1,146 @@
+import { Writable } from 'node:stream'
+import path from 'path'
+
+import React from 'react'
+
+import { renderToPipeableStream, renderToString } from 'react-dom/server'
+
+import { getPaths } from '@redwoodjs/project-config'
+import type { TagDescriptor } from '@redwoodjs/web'
+// @TODO (ESM), use exports field. Cannot import from web because of index exports
+import {
+  ServerHtmlProvider,
+  ServerInjectedHtml,
+  createInjector,
+  RenderCallback,
+} from '@redwoodjs/web/dist/components/ServerInject'
+
+interface RenderToStreamArgs {
+  ServerEntry: any
+  currentPathName: string
+  metaTags: TagDescriptor[]
+  includeJs: boolean
+  res: Writable
+}
+
+export function reactRenderToStream({
+  ServerEntry,
+  currentPathName,
+  metaTags,
+  includeJs,
+  res,
+}: RenderToStreamArgs) {
+  const rwPaths = getPaths()
+
+  const bootstrapModules = [
+    path.join(__dirname, '../inject', 'reactRefresh.js'),
+  ]
+
+  if (includeJs) {
+    // type casting: guaranteed to have entryClient by this stage, because checks run earlier
+    bootstrapModules.push(rwPaths.web.entryClient as string)
+  }
+
+  // TODO (STREAMING) CSS is handled by Vite in dev mode, we don't need to
+  // worry about it in dev but..... it causes a flash of unstyled content.
+  // For now I'm just injecting index css here
+  // Looks at collectStyles in packages/vite/src/fully-react/find-styles.ts
+  const FIXME_HardcodedIndexCss = ['index.css']
+
+  const assetMap = JSON.stringify({
+    css: FIXME_HardcodedIndexCss,
+    meta: metaTags,
+  })
+
+  // This ensures an isolated state for each request
+  const { injectionState, injectToPage } = createInjector()
+
+  // This is effectively a transformer stream
+  const intermediateStream = createServerInjectionStream({
+    outputStream: res,
+    onFinal: () => {
+      res.end()
+    },
+    injectionState,
+  })
+
+  const { pipe } = renderToPipeableStream(
+    React.createElement(
+      ServerHtmlProvider,
+      {
+        value: injectToPage,
+      },
+      ServerEntry({
+        url: currentPathName,
+        css: FIXME_HardcodedIndexCss,
+        meta: metaTags,
+      })
+    ),
+    {
+      bootstrapScriptContent: includeJs
+        ? `window.__assetMap = function() { return ${assetMap} }`
```

Comment:
Not sure what other code would need to change, but maybe this should be prefixed with `__REDWOOD__` like the helmet provider? Just to distinguish it as coming from our code. Not sure how the caps would work, just a suggestion to run with:

```suggestion
        ? `window.__REDWOOD__assetMap = function() { return ${assetMap} }`
```

Reply:
You're right... i plan on changing the implementation here so they might not exist later, but I'll do these as well for now.